File: python/semantic/big_step.py
# coding: utf-8


class Number(object):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return str(self.value)

    def eval(self, env):
        return self


class Add(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return '{0} + {1}'.format(self.left, self.right)

    def eval(self, env):
        return Number(self.left.eval(env).value + self.right.eval(env).value)


class Multiply(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return '{0} * {1}'.format(self.left, self.right)

    def eval(self, env):
        return Number(self.left.eval(env).value * self.right.eval(env).value)


class Boolean(object):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return str(self.value)

    def eval(self):
        return self


class LessThan(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return '{0} < {1}'.format(self.left, self.right)

    def eval(self, env):
        return Boolean(self.left.eval(env).value < self.right.eval(env).value)


class Variable(object):
    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return str(self.name)

    def eval(self, env):
        return env[self.name]


class DoNothing(object):
    def __eq__(self, other_statement):
        return isinstance(other_statement, DoNothing)

    def __repr__(self):
        return 'do_nothing'

    def eval(self, env):
        return env


class Assign(object):
    def __init__(self, name, expr):
        self.name = name
        self.expr = expr

    def __repr__(self):
        return '{0} = {1}'.format(self.name, self.expr)

    def eval(self, env):
        env.update({self.name: self.expr.eval(env)})
        return env


class If(object):
    def __init__(self, condition, sequence, alternative):
        self.condition = condition
        self.sequence = sequence
        self.alternative = alternative

    def __repr__(self):
        return 'If ({0}) {{ {1} }} else {{ {2} }}'.format(
            self.condition, self.sequence, self.alternative
        )

    def eval(self, env):
        if self.condition.eval(env).value == Boolean(True).value:
            return self.sequence.eval(env)
        return self.alternative.eval(env)


class Sequence(object):
    def __init__(self, first, second):
        self.first = first
        self.second = second

    def __repr__(self):
        return '{0}; {1}'.format(self.first, self.second)

    def eval(self, env):
        return self.second.eval(self.first.eval(env))


class While(object):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body

    def __repr__(self):
        return 'while ({0}) {{ {1} }}'.format(self.condition, self.body)

    def eval(self, env):
        if self.condition.eval(env).value == Boolean(True).value:
            return self.eval(self.body.eval(env))
        return env


# Test "Number"
print Number(23).eval({})

# Test "Variable"
print Variable('x').eval({'x': Number(3)})


# Test "Add", "LessThan"
print LessThan(Add(Variable('x'), Number(2)),
               Variable('y')
               ).eval({'x': Number(1), 'y': Number(5)})

# Test "Sequence"
print Sequence(Assign('y', Number(1)),
                  Assign('z', Number(2))).eval({})

# Test "If"
print If(Variable('x'),
            Sequence(Assign('y', Number(1)),
                         Assign('z', Number(2))),
            DoNothing
            ).eval({'x': Boolean(True)})

# Test "While"
print While(LessThan(Variable('x'), Number(6)),
               Assign('x', Multiply(Variable('x'), Number(2)))
               ).eval({'x': Number(1)})


## Test "If", "Sequence", "Assign"
# statement = If(Variable('x'),
#                Sequence(Assign('y', Number(1)),
#                         Assign('z', Number(2))),
#                DoNothing())
# env = {'x': Boolean(True)}

## Test "While"
# statement = While(LessThan(Variable('x'), Number(6)),
#                   Assign('x', Multiply(Variable('x'), Number(2))))
# env = {'x': Number(1)}

File: ruby/semantic/big_step.rb


class Number < Struct.new(:value)
  def to_s
    "#{value}"
  end

  def inspect
    "<<#{self}>>"
  end

  def eval(env)
    self
  end
end


class Add < Struct.new(:left, :right)
  def to_s
    "#{left}+#{right}"
  end

  def eval(env)
    Number.new(left.eval(env).value + right.eval(env).value)
  end
end


class Multiply < Struct.new(:left, :right)
  def to_s
    "#{left}*#{right}"
  end

  def eval(env)
    Number.new(left.eval(env).value * right.eval(env).value)
  end
end


class Boolean < Struct.new(:value)
  def to_s
    value.to_s
  end

  def inspect
    "<<#{self}>>"
  end

  def eval(env)
    self
  end
end


class LessThan < Struct.new(:left, :right)
  def to_s
    "#{left} < #{right}"
  end

  def eval(env)
    Boolean.new(left.eval(env).value < right.eval(env).value)
  end
end


class Variable < Struct.new(:name)
  def to_s
    name.to_s
  end

  def eval(env)
    env[name]
  end
end


class DoNothing
  def to_s
    'do_nothing'
  end

  def eval(env)
    env
  end
end


class Assign < Struct.new(:name, :expr)
  def to_s
    "#{name} = #{expr}"
  end

  def eval(env)
    env.merge({name => expr.eval(env)})
  end
end


class If < Struct.new(:condition, :consequence, :alternative)
  def to_s
    "if (#{condition}) { #{consequence} } else { #{alternative} }"
  end

  def eval(env)
    case condition.eval(env)
      when Boolean.new(true)
        consequence.eval(env)
      when Boolean.new(false)
        alternative.eval(env)
    end
  end
end


class Sequence < Struct.new(:first, :second)
  def to_s
    "#{first}; #{second}"
  end

  def eval(env)
    second.eval(first.eval(env))
  end
end


class While < Struct.new(:condition, :body)
  def to_s
    "while (#{condition}) { #{body} }"
  end

  def eval(env)
    case condition.eval(env)
      when Boolean.new(true)
        eval(body.eval(env))
      when Boolean.new(false)
        env
    end
  end
end


# # Test "Number"
# puts Number.new(23).eval({})
#
# # Test "Variable"
# puts Variable.new(:x).eval({x: Number.new(3)})
#
# # Test "Add", "LessThan"
# puts LessThan.new(Add.new(Variable.new(:x), Number.new(2)),
#                   Variable.new(:y)
#                   ).eval({x: Number.new(1), y: Number.new(5)})
#
# # Test "Sequence"
# puts Sequence.new(Assign.new(:y, Number.new(1)),
#                   Assign.new(:z, Number.new(2))).eval({})
#
# # Test "If"
# puts If.new(Variable.new(:x),
#             Sequence.new(Assign.new(:y, Number.new(1)),
#                          Assign.new(:z, Number.new(2))),
#             DoNothing.new
#             ).eval({x: Boolean.new(true)})
#
# # Test "While"
# puts While.new(LessThan.new(Variable.new(:x), Number.new(6)),
#                Assign.new(:x, Multiply.new(Variable.new(:x), Number.new(2)))
#                ).eval({x: Number.new(1)})
